### ISSUE TITLE
False positive in `needless_range_loop`

### ISSUE THREAD

**Issue by chyvonomys** [rust-lang-nursery/rust-clippy#2036]
Clippy reports `needless_range_loop` and suggests using iter/take/skip which is wrong for this case:
```
let ns = [2, 3, 5, 7];
for i in 3..10 {
    println!("{}", ns[i % 4]);
}
```
```
warning: the loop variable `i` is only used to index `ns`.
 --> src\main.rs:3:5
  |
3 | /     for i in 3..10 {
4 | |         println!("{}", ns[i % 4]);
5 | |     }
  | |_____^
  |
  = note: #[warn(needless_range_loop)] on by default
  = help: for further information visit https://rust-lang-nursery.github.io/rust-clippy/v0.0.158/index.html#needless_range_loop
help: consider using an iterator
  |
3 |     for <item> in ns.iter().take(10).skip(3) {
  |         ^^^^^^
```

**Comment by chyvonomys** [comment]
Looks like this is related to this piece:
https://github.com/rust-lang-nursery/rust-clippy/blob/9c9a4953c3f131a8481f15f7e7a805972056a084/clippy_lints/src/loops.rs#L1346-L1354

While `same_var(...)` will correctly catch cases like: `ns[i]`, second case catch things like `ns[i % 4]`.

Code was introduced with this PR: https://github.com/rust-lang-nursery/rust-clippy/pull/2021

**Comment by oli-obk** [comment]
Jup, this should probably suggest `ns[3..7].iter().cycle().take(10)`

**Comment by chyvonomys** [comment]
I think it would be better to make it work as before, i.e. to report cases like `ns[i]` only. Expression under `[...]` may be more complex than just `i % 4` and figuring out what iterator style equivalent to suggest seems to be a non-trivial task. Also, using just expression is better than iterator. At least in my example.

**Comment by oli-obk** [comment]
fixed in https://github.com/rust-lang-nursery/rust-clippy/pull/2118
